main.go

```diff
@@ -73,60 +68,41 @@ func testUSBFunctionalityAvailableOnThisPlatform() (err error) {
-For USB devices, use something like: {"usb": [{"vendor": "1209", "product": "000F"}]}
-For example, to expose serial devices with different names: {"name": "serial", "groups": [{"paths": [{"path": "/dev/ttyUSB*"}]}, {"paths": [{"path": "/dev/ttyACM*"}]}]}
-The device flag can specify lists of devices that should be grouped and mounted into a container together as one single meta-device.
-For example, to allocate and mount an audio capture device: {"name": "capture", "groups": [{"paths": [{"path": "/dev/snd/pcmC0D0c"}, {"path": "/dev/snd/controlC0"}]}]}
-For example, to expose a CH340 serial converter: {"name": "ch340", "groups": [{"usb": [{"vendor": "1a86", "product": "7523"}]}]}
-A "count" can be specified to allow a discovered device group to be scheduled multiple times.
-For example, to permit allocation of the FUSE device 10 times: {"name": "fuse", "groups": [{"count": 10, "paths": [{"path": "/dev/fuse"}]}]}
-Note: if omitted, "count" is assumed to be 1`)
-	pluginPath := flag.String("plugin-directory", v1beta1.DevicePluginPath, "The directory in which to create plugin sockets.")
-	logLevel := flag.String("log-level", logLevelInfo, fmt.Sprintf("Log level to use. Possible values: %s", availableLogLevels))
-	listen := flag.String("listen", ":8080", "The address at which to listen for health and metrics.")
-	printVersion := flag.Bool("version", false, "Print version and exit")
-	flag.Parse()
-
-	if *printVersion {
+	if err := InitConfig(); err != nil {
+		return err
+	}
+
+	if viper.GetBool("version") {
 		fmt.Println(version.Version)
 		return nil
 	}
 
-	if errs := validation.IsDNS1123Subdomain(*domain); len(errs) > 0 {
-		return fmt.Errorf("failed to parse domain %q: %s", *domain, strings.Join(errs, ", "))
+	domain := viper.GetString("domain")
+	if errs := validation.IsDNS1123Subdomain(domain); len(errs) > 0 {
+		return fmt.Errorf("failed to parse domain %q: %s", domain, strings.Join(errs, ", "))
 	}
 
 	deviceTypeFmt := "[a-z0-9][-a-z0-9]*[a-z0-9]"
 	deviceTypeRegexp := regexp.MustCompile("^" + deviceTypeFmt + "$")
 	var trim string
-	deviceSpecs := make([]*deviceplugin.DeviceSpec, len(*deviceSpecsRaw))
 	var shouldTestUSBAvailable bool
-	for i, dsr := range *deviceSpecsRaw {
-		if err := json.Unmarshal([]byte(dsr), &deviceSpecs[i]); err != nil {
-			return fmt.Errorf(
-				"failed to parse device %q; see --help",
-				dsr,
-			)
-		}
+	deviceSpecs, err := GetDevices()
+	if err != nil {
+		return err
+	}
+	for i, dsr := range deviceSpecs {
 		// Apply defaults.
 		deviceSpecs[i].Default()
 		trim = strings.TrimSpace(deviceSpecs[i].Name)
 		if !deviceTypeRegexp.MatchString(trim) {
-			return fmt.Errorf("failed to parse device %q; device type must match the regular expression %q", dsr, deviceTypeFmt)
+			return fmt.Errorf("failed to parse device %s; device type must match the regular expression %q", dsr.Name, deviceTypeFmt)
```

Comment:
Keeping this quoted is helpful IMO to show that it's variable, user content and not a hardcoded string from the code.
```suggestion
			return fmt.Errorf("failed to parse device %q; device type must match the regular expression %q", dsr.Name, deviceTypeFmt)
```